### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by ridams** [abiliojr/fts5-snowball#3]
Hi, this is a nice and clean code, however, I couldnt make it work in multithreaded environnement, also how should we do to make directly available just like fts5

**Comment by and-se** [comment]
I faced the same problem. Some investigations led to the following.
This extension has global snowball stemmers cache as `struct StemmerListItem *availableStemmers`.

It is not thread safe, because:
1. stemmer object is not thread safe - `libstemmer.h` says, that next call to `sb_stemmer_stem` invalidates previous token info (and so `sb_stemmer_length`). So we have race condition when many threads work with the same stemmer.
2. `destroySnowball` handler called on db connection close is not thread save too. Under heavy load we can get two connections in `destroySnowball` handler, so we will free cache twice.

This problem can be fixed so:
* use locks:
  * lock stemmer untill all necessarry data copied somewhere.
  * lock when destroy cache
  * :heavy_plus_sign: Single stemmers cache - save ram (single cache) and cpu (one stemmer instance creation per language)
  * :heavy_minus_sign: working with stemmers is bottleneck due to locks.
  * :information_source: we can try use portable [sqlite mutexes](https://www.sqlite.org/c3ref/mutex_alloc.html)
* make cache thread_local
  * :heavy_minus_sign: more ram and cpu consumed, because each thread have own stemmer instance
  * :heavy_plus_sign: no bottleneck with locks
  * :heavy_plus_sign: simple to implement - just write something like `__thread`
  * :information_source: thread_local supported by compilers - we can try [this macros](https://stackoverflow.com/questions/18298280/how-to-declare-a-variable-as-thread-local-portably)
* remove cache at all:
  * :heavy_minus_sign: slightly difficult to implement - many code depends on `availableStemmers` indexes.
  * :heavy_minus_sign: cpu and ram consumption
  * :heavy_plus_sign: simple - nothing to be broken...

Now I try use thread_local. It immediately helps with race condition on stemmers, but not helps with destroySnowball - we still have race condition between two db connections in **one thread**:
* conn1 closes and disposes cache `availableStemmers=NULL`
* conn2 calls `fts5SnowballCb` and fails when try do `availableStemmers[...]`, because it is `NULL`.
It can be fixed by something like ref_counter - how many connections in the same thread use this cache. This idea is applicable for locks-solution too - why destroy global cache if it needed for other connections?


If I am successful, I will send a PR later.

**Comment by and-se** [comment]
"Race condition" in one thread with `destroySnowball` may occur implicitly in web api app if:
* you have gc language (python)
* every request processed in own thread (fastapi `def` routes)
* thread got from thread pool
* you forgot eplicitly close db connection in python code

I discovered this scenario under load:
* thread1 processes request1 - opens conn1 to db
* request finished, but no explicit conn1.close()
* thread1 is free in thread pool
* other request2 get thread1 and processes request - opens another conn2
* python gc wakes up and closes conn1 - so call `destroySnowball`, so `availableStemmers=NULL`
* request2 tries to continue work with conn2, but it has broken `availableStemmers`! seg fault in `fts5SnowballCb` on line
 `stemmed = (sb_symbol *) sb_stemmer_stem(availableStemmers[*stemmers].stemmer, (unsigned char*) aBuf, originalNBuf);`.


Maybe I made a mistake and something else is happening, but this hypothesis is supported by the following techniques:
* disable `destroySnowball` code - all works!
*  in `fts5SnowballCb`  check `availableStemmers=NULL` and `printf("NULL!")`. It becomes last message before segfault.

**Comment by and-se** [comment]
Ref count idea + thread local stemmers cache have a problem: it will not work as expected in python if:
* multithread app, each thread opens db connection, loads this ext, do some work, return thread to thread pool.
* no explicit db close again...
* so GC goes to close db connections. But it works in a single thread, not where the db connection was created!

So we have memory leak. because thread local cache is not freed.

Of course, we can require the python programmer to close all connections explicitly, but...

**Comment by abiliojr** [comment]
@and-se, thanks for your detailed analysis. I'm away from my pc this weekend, but I will try to take a look at it next week. You mentioned pytho: do you have a python minimum working example?

**Comment by and-se** [comment]
Hello! The original test uses JMeter and a big python application (which uses fastapi), but it seems I managed to make a toy script that crashes in about the same way

```python
import sqlite3
from threading import get_ident as thread_id
from concurrent.futures import ThreadPoolExecutor
import random

def connect():
    c = sqlite3.connect('testdb.sqlite3')
    c.enable_load_extension(True)
    c.load_extension("./fts5stemmer")
    return c
    
c = connect()
c.executescript("""
create virtual table if not exists Test using fts5(text, tokenize='snowball');
insert into Test values ("cats dogs"), ("cats dogs"), ("horse dogs")
""")  # I remember MIN_TOKEN_LENGTH = 3

sh = 10000  # cut ids

class ConnWrap:
    def __init__(self, conn):
        self.conn = conn
        self.thread = thread_id()
        self.cycle = self  # cyclic reference to avoid gc remove object immediately
    
    def execute(self, sql, params):
        return self.conn.execute(sql, params)
    
    def executescript(self, sql):
        return self.conn.executescript(sql)
        
    def __del__(self):
        if thread_id() != self.thread:
            # in thread_local cache  we need to delete the connection in the same thread where it was created - else memory leak
            print(f"!!! thread={thread_id()%sh} GC delete conn {id(self)%sh} - expected in thread={self.thread%sh}!")
            
def do_select(conn):
    conn.execute("select * from Test where Test = ?", 
        (random.choice("cats dogs horse".split()),)).fetchall() 

def tester(job, test_inner=True):
    """
    test_inner=True - test seg fault when two connections in one thread
    modify to test_inner=False to catch another seg fault
    """
    c = ConnWrap(connect())
    title=f"thread={thread_id()%sh} job={job%sh} conn={id(c)%sh}"
    print(title, "started")
    do_select(c)
    
    if test_inner:
        # seg fault two connections in one thread
        print(title, "inner conn start")
        c2 = connect()
        do_select(c2)
        c2.close()    
        print(title, "inner conn end")
    
    do_select(c)
    # c.close() # no explicit db close - allow gc do work (in wrong thread!)
    print(title, "done")

# play around with this parameters
# my python 3.7.5 always fails with test_inner=True
# if test_inner=False, often fails with (20, 100)
with ThreadPoolExecutor(2) as executor:
    executor.map(tester, range(10))
    
# modify fts5stemmer.c lines 21-22:
# __thread struct StemmerListItem *availableStemmers;
# __thread int numberAvailableStemmers = 0;
# 
# when test_inner=False segfault stops, but GC delete problems remain.
# when test_inner=True segfault continues.


print("All OK. May be...")

```

**Comment by and-se** [comment]
I have solution - see PR #9. It works until we have a separate connection for each thread (i believe this is usually the case).
It's tested with my real app under JMeter under heavy load - no seg faults.

It works with python script above. Warnings about deleting a connection from another thread are still issued, but this is not a problem for this solution.

**Comment by abiliojr** [comment]
@and-se, @ridams I have a different proposal that seems to work (at least with the provided python script). Please see PR #10. I like it because it simplifies the code a lot (please see commit 553fef2b21227cc5b746804b59c53512c6056538). Would you be so kind to test it in your application and confirm it works? If you have any comments, please let me know.

**Comment by and-se** [comment]
Thank you! This is a really nice solution.

I have tested in in my app - works fine.
There is no cache in your solution, but I could not detect a noticeable performance drop compared to my solution.

If I have any problems, I'll let you know, but now you can merge it.

**Comment by abiliojr** [comment]
Thanks for the confirmation. Merged #11, so I'm closing this issue
